charmcraft/commands/build.py

```diff
@@ -0,0 +1,178 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# For further info, check https://github.com/canonical/charmcraft
+
+
+import logging
+import os
+import pathlib
+import shutil
+import stat
+import subprocess
+import zipfile
+
+from charmcraft.cmdbase import BaseCommand, CommandError
+
+logger = logging.getLogger(__name__)
+
+# Some constants that are used through the code.
+CHARM_METADATA = 'metadata.yaml'
+BUILD_DIRNAME = 'build'
+DISPATCH_FILENAME = 'dispatch'
+
+# The template for the dispatch script
+DISPATCH = """#!/bin/sh
+
+PYTHONPATH=lib {entrypoint_relative_path}
+"""
+
+# The minimum set of hooks to be provided for compatibility with old Juju
+HOOK_NAMES = [
+    'install',
+    'start',
+    'upgrade-charm',
+]
+
+
+def polite_exec(cmd):
+    """Execute a command, only showing output if error."""
+    p = subprocess.Popen(
+        cmd, stdout=subprocess.PIPE, stderr=subprocess.STDOUT, universal_newlines=True)
+
+    stdout = []
+    for line in p.stdout:
+        stdout.append(line)
+    retcode = p.wait()
+
+    if retcode:
+        logger.error("Execution ended in %d for cmd %s", retcode, cmd)
+        for line in stdout:
+            logger.debug(":: %s", line.rstrip())
+    return retcode
+
+
+def link(srcpath, destdir):
+    """Link a file."""
+    if not destdir.exists():
+        os.makedirs(destdir)
+
+    # XXX Facundo 2020-05-19: we could make hard links here, with the benefit of `juju deploy`
+    # being happy about these links, while at the same time getting the benefit of "changing
+    # a file in the project and being ready to deploy" (however, TRICKY!); that said, we would
+    # need to handle directories by hand (but currently we're not linking dirs...)
+    destpath = destdir / srcpath.name
+    destpath.symlink_to(srcpath)
+    return destpath
+
+
+def _build(charmdir, entrypoint):
+    """Main entry point."""
+    buildpath = charmdir / BUILD_DIRNAME
+    logger.debug("Building charm in %r", str(buildpath))
+
+    # XXX Facundo 2020-05-18: for now we're re-building always from scratch, in the
+    # future we *may* reuse stuff already inside
+    if buildpath.exists():
+        shutil.rmtree(str(buildpath))
+    buildpath.mkdir()
+
+    # basic files
+    logger.debug("Linking in basic files and charm code")
+    link(charmdir / CHARM_METADATA, buildpath)
+
+    # the charm code
+    # XXX Facundo 2020-05-18: for now, one file, may we copy the entrypoint and its whole dir/tree
+    # (if not charmdir); also we need to understand if we want to ignore some files (e.g. .pyc)
+    linked_entrypoint = link(entrypoint, buildpath / 'src')
+
+    # dispatch mechanism
+    logger.debug("Creating the dispatch mechanism")
+    dispatch_content = DISPATCH.format(
+        entrypoint_relative_path=linked_entrypoint.relative_to(buildpath))
+    dispatch_path = buildpath / DISPATCH_FILENAME
+    with open(dispatch_path, "wt", encoding="utf8") as fh:
```

Comment:
its a shame that os.open() supports a mask but that isn't exposed in any way for open().
I don't think it matters much here, though. Usually the open-with-mask is because you don't want to create the file readable by others before you write to it, because then other people can spin and grab the file handle before you chmod it.

Reply:
Yes, I improved it.

Reply:
"best" is subjective of course, and if it were particularly sensitive I'd suggest using os.open directly :-) (`open()` can take a fd instead of a path)